internal/se/se.go

```diff
@@ -0,0 +1,212 @@
+// and returns the document IDs for documents that match all queries.
+func fetchDocIDs(
+	ctx context.Context,
+	ds corekv.ReaderWriter,
+	collectionID string,
+	queries []FieldQuery,
+) ([]string, error) {
+	docIDSet := make(map[string]struct{})
+
+	isFirstPass := true
+	for _, query := range queries {
+		key := keys.DatastoreSE{
+			CollectionID: collectionID,
+			IndexID:      query.IndexID,
+			SearchTag:    query.SearchTag,
+		}
+
+		iter, err := ds.Iterator(ctx, corekv.IterOptions{
+			Prefix: key.Bytes(),
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		querySet := make(map[string]struct{})
+		for {
+			hasNext, err := iter.Next()
+			if err != nil || !hasNext {
+				break
+			}
+
+			dsKey, err := keys.NewDatastoreSEFromString(string(iter.Key()))
+			if err != nil {
+				return nil, errors.Join(err, iter.Close())
+			}
+			if dsKey.DocID == "" {
+				return nil, errors.Join(NewErrEmptyDocID(dsKey.ToString()), iter.Close())
+			}
+
+			querySet[dsKey.DocID] = struct{}{}
+		}
+
+		err = iter.Close()
+		if err != nil {
+			return nil, err
+		}
+
+		if isFirstPass {
+			docIDSet = querySet
+			isFirstPass = false
+		} else {
+			for docID := range docIDSet {
+				if _, exists := querySet[docID]; !exists {
+					delete(docIDSet, docID)
+				}
+			}
+		}
+
+		if len(docIDSet) == 0 {
+			break
+		}
+	}
+
+	docIDs := make([]string, 0, len(docIDSet))
+	for docID := range docIDSet {
+		docIDs = append(docIDs, docID)
+	}
+
+	return docIDs, nil
+}
+
+// FieldQuery represents a query for a specific encrypted field
+type FieldQuery struct {
+	FieldName string
+	IndexID   string
+	SearchTag []byte
+}
+
+// generateDocArtifacts generates SE artifacts for specified fields of a document.
+// If fieldNames is empty or nil, artifacts are generated for all encrypted fields.
+func generateDocArtifacts(
+	ctx context.Context,
+	col client.Collection,
+	doc *client.Document,
+	fieldNames []string,
+	encKey []byte,
+) ([]secore.Artifact, error) {
+	encryptedIndexes, err := col.ListEncryptedIndexes(ctx)
+	if err != nil {
+		return nil, NewErrFailedToGetEncryptedIndexes(err)
+	}
+
+	if len(encryptedIndexes) == 0 {
+		return nil, nil
+	}
+
+	collectionID := col.VersionID()
+	docID := doc.ID().String()
+
+	var artifacts []secore.Artifact
+	for _, encIdx := range encryptedIndexes {
+		// Skip if fieldNames is specified and this field is not in the list
+		if len(fieldNames) > 0 && !slices.Contains(fieldNames, encIdx.FieldName) {
+			continue
+		}
+
+		fieldValue, err := doc.GetValue(encIdx.FieldName)
+		if err != nil {
+			return nil, NewErrFailedToGetFieldValue(encIdx.FieldName, err)
+		}
+
+		normalValue := fieldValue.NormalValue()
+		artifact, err := GenerateFieldArtifact(ctx, collectionID, docID, encIdx, normalValue, encKey)
+		if err != nil {
+			return nil, err
+		}
+		artifacts = append(artifacts, artifact)
+	}
+
+	return artifacts, nil
+}
+
+// GenerateFieldArtifact generates a single SE artifact for a specific field value.
+func GenerateFieldArtifact(
```

Comment:
question: This looks like it is only called from within this package, but I might still have some collapsed files - why is it public?